arbitrator/prover/src/merkle.rs

```diff
@@ -43,81 +131,200 @@ impl MerkleType {
+        let mut layer_i = 0usize;
         while layers.last().unwrap().len() > 1 || layers.len() < min_depth {
-            let empty_layer = *empty_layers.last().unwrap();
-
-            #[cfg(feature = "rayon")]
-            let new_layer = layers.last().unwrap().par_chunks(2);
+            let layer = layers.last().unwrap();
+            let empty_hash = empty_hash_at(ty, layer_i);
 
-            #[cfg(not(feature = "rayon"))]
-            let new_layer = layers.last().unwrap().chunks(2);
-
-            let new_layer = new_layer
-                .map(|chunk| hash_node(ty, chunk[0], chunk.get(1).cloned().unwrap_or(empty_layer)))
-                .collect();
-            empty_layers.push(hash_node(ty, empty_layer, empty_layer));
+            let new_layer = new_layer(ty, layer, empty_hash);
+            dirty_indices.push(HashSet::with_capacity(new_layer.len()));
             layers.push(new_layer);
+            layer_i += 1;
         }
+        let layers = Mutex::new(Layers {
+            data: layers,
+            dirt: dirty_indices,
+        });
         Merkle {
             ty,
             layers,
-            empty_layers,
             min_depth,
         }
     }
 
+    fn rehash(&self, layers: &mut Layers) {
+        // If nothing is dirty, then there's no need to rehash.
+        if layers.dirt.is_empty() || layers.dirt[0].is_empty() {
+            return;
+        }
+        // Process dirty indices starting from layer 1 (layer 0 is the leaves).
+        for layer_i in 1..layers.data.len() {
+            let dirty_i = layer_i - 1;
+            // Consume this layer's dirty indices.
+            let dirt = std::mem::take(&mut layers.dirt[dirty_i]);
+            // It is important to process the dirty indices in order because
+            // when the leaves grown since the last rehash, the new parent is
+            // simply pused to the end of the layer's data.
+            for idx in dirt.iter().sorted() {
+                let left_child_idx = idx << 1;
+                let right_child_idx = left_child_idx + 1;
+                // The left child is guaranteed to exist, but the right one
+                // might not if the number of child nodes is odd.
+                let left = layers.data[layer_i - 1][left_child_idx];
+                let right = layers.data[layer_i - 1]
+                    .get(right_child_idx)
+                    .unwrap_or(empty_hash_at(self.ty, layer_i - 1));
+                let new_hash = hash_node(self.ty, left, right);
+                if *idx < layers.data[layer_i].len() {
+                    layers.data[layer_i][*idx] = new_hash;
+                } else {
+                    // Push the new parent hash onto the end of the layer.
+                    layers.data[layer_i].push(new_hash);
+                }
+                // Mark the node's parent as dirty unless it's the root.
+                if layer_i < layers.data.len() - 1 {
+                    layers.dirt[dirty_i + 1].insert(idx >> 1);
```

Comment:
One thing that just occurred to me, do we actually need to store the parent node's dirty status in this struct? Aren't we always just going to clear it out as we iterate over the next layer? I think we could only store the dirty leaves, and then in this function keep track of the next layer's dirty indices locally.

Reply:
One trick you could theoretically do is to keep:
- the dirty you're currently iterating over
- the next dirty
and then when you're done, swap them, slice the next dirty down to the new size, and set it all to zero.

That said, I don't think it's worth implementing given the results of your benchmark.